kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,591 @@
+    if not errcode then
+        local ttl = answers[1].ttl
+
+        for _, answer in ipairs(answers) do
+            -- A compromise regarding https://github.com/Kong/kong/pull/3088
+            if answer.type == TYPE_AAAA then
+                answer.address = utils.ipv6_bracket(answer.address)
+            elseif answer.type == TYPE_SRV then
+                answer.target = utils.ipv6_bracket(answer.target)
+            end
+
+            ttl = math_min(ttl, answer.ttl)
+        end
+
+        answers.ttl = self.valid_ttl or ttl
+
+    elseif errcode == 3 or errcode == 101 then
+        answers.ttl = self.empty_ttl
+
+    else
+        answers.ttl = self.error_ttl
+    end
+end
+
+
+-- NOTE: it might insert unmatched answers into cache
+local function process_answers(self, qname, qtype, answers)
+    if not answers.errcode then
+        local unmatched = filter_unmatched_answers(qname, qtype, answers)
+        for k, a in pairs(unmatched) do
+            process_answers_fields(self, a)
+            self.cache:set(k, { ttl = a.ttl }, a)
+            if not get_last_type(self.cache, a[1].name) then
+                insert_last_type(self.cache, a[1].name, a[1].type)
+            end
+        end
+
+        if #answers == 0 then
+            answers.errcode = 101
+            answers.errstr = client_errors[101]
+        end
+    end
+
+    process_answers_fields(self, answers)
+end
+
+
+local function resolve_query(self, name, qtype, tries)
+    --log(tries, "query")
+
+    local key = name .. ":" .. qtype
+    stats_count(self.stats, key, "query")
+
+    local r, err = resolver:new(self.r_opts)
+    if not r then
+        return nil, "failed to instantiate the resolver: " .. err
+    end
+
+    local options = { additional_section = true, qtype = qtype }
+    local answers, err, q_tries = r:query(name, options, {})
+    if r.destroy then
+        r:destroy()
+    end
+
+    if not answers then
+        stats_count(self.stats, key, "query_fail")
+        return nil, "Query to DNS server failed: " .. (err or "unknown")
+    end
+
+    process_answers(self, name, qtype, answers)
+
+    stats_count(self.stats, key, answers.errstr and
+                                 "query_err:" .. answers.errstr or "query_succ")
+
+    --log(tries, answers.errstr or #answers)
+
+    return answers, nil, answers.ttl
+end
+
+
+local function start_stale_update_task(self, key, name, qtype)
+    stats_count(self.stats, key, "stale")
+
+    timer_at(0, function (premature)
+        if premature then return end
+
+        local answers = resolve_query(self, name, qtype, {})
+        if answers and (not answers.errcode or answers.errcode == 3) then
+            self.cache:set(key, { ttl = answers.ttl }, answers)
+            insert_last_type(self.cache, name, qtype)
+        end
+    end)
+end
+
+
+local function resolve_name_type_callback(self, name, qtype, opts, tries)
+    local key = name .. ":" .. qtype
+
+    local ttl, err, answers = self.cache:peek(key, true)
```

Comment:
I am not sure what this is trying to do, but using `peek()` from inside an L3 callback value is extremely suspicious to me and looks like an anti-pattern.

Reply:
Here I use `:peek()` to check whether the key is expired but still present in the L2 shared dictionary. If it is expired but still present, it will initiate an asynchronous background updating task to refresh the outdated record.

I do not use mlcache option `resurrect_ttl` to implement the stale logic, because:
1.  Callers of ":get()" will be block by the network query of L3 callback although it could get the stale data from the cache. Here I want to implement an asynchronous background updating mechanism. 
2. And if the L3 callback consistently fails, it will indefinitely use the stale data. Here, I would like to make the expired data only available within the `dns_client:stale_ttl` period, without continuously refreshing its expiration time.